### PR TITLE
Feat: implement addon init command

### LINKED ISSUE #4116: [Feature] support archive addon as a helm chart that can be installed by `helm install`
**Is your feature request related to a problem? Please describe.**

We must use `vela addon enable` to install an addon, while it seems everything can be done by using `helm install` ?

We add  burden to end user that they must know the concept of addon.

**Describe the solution you'd like**

1. Add a `vela addon archive` command that can convert current addon to a helm chart, which can be uploaded to the chart museum.
2. The helm chart addon will deploy by `helm install` into the control plane cluster. It leverages application controller for multi-cluster delivery just like what it works now. Anything that only need to be installed in the control-plane should just be yaml files in the chart while resources need to be delivered into multi-clusters should be in the application yaml.
3. The rest thing can be aligned with helm chart, such as Notes.txt, parameters.cue(values.yaml), etc.

**Describe alternatives you've considered**

The currunt solution use `vela addon` command for installation, actually we are building a new wheel which is no better.

**Additional context**

Hope there's some unique features that make `vela addon enable` a necessary thing. In this case, `vela addon archive` is necessary as it converts cue into helm charts. \cc @wangyikewxgm 

## What makes this issue?

* #4114 seems just a Notes.txt in helm chart but we're rebuilding in the addon system.
* #3822 helm chart already solves this by it's go templating 
* #3905 we wrap all the things in the addon and deploy them without thinking, if addon align with helm chart and it contains an application inside, everything become clear.
* #3891  helm chart also has this issue with a solution that uses label check.
* #3692 same issue with helm

We MUST solve this before any progress of addon framework.

## What Values Addon should ADD?

Any feature as a package management tool, the addon should be aligned with Helm. So addon can hardly add values besides:

1. Automation: convert a helm chart into an addon by automatically detect the CRD controller and add OAM Definition for it.
2. Glue/Orchestration: addon can provide an abstraction layer for the real Implementation, which means one addon can depends on another addon which can be different implementation( different helm charts).

### PR DESCRIPTION
**Pull request by charlie0129**
### Description of your changes

As described in https://github.com/kubevela/kubevela/issues/4116#issuecomment-1148450959, this pr adds a `vela addon init` command.

- If the user provides all three Chart-related parameters, an addon scaffold, with a `helm` component inside, will be created.
- If the user does't provide all the necessary parameters to create a `helm` component, a basic addon scaffold is generated instead.

This pr does **not** close that issue, since more features are needed.

<details>
<summary>Command Help Text</summary>

```console
$ vela addon init -h
Create an addon scaffold for quick starting. A Helm Component is generated if you provide Chart-related parameters.

Usage:
  vela addon init [flags]

Examples:
	vela addon init mongodb --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16
will create something like this:
	mongodb/
	├── definitions
	├── metadata.yaml
	├── readme.md
	├── resources
	│   └── mongodb.cue
	├── schemas
	└── template.yaml

If you want to store the scaffold in a different directory, you can use the -p/--path flag:
	vela addon init mongodb -p ./some/repo --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16

If you don't want the Helm component, just omit the three Chart-related parameters. We will create an empty scaffold for you.
	vela addon init mongodb

Flags:
      --chart string           Helm Chart name
      --helm-repo-url string   URL that points to a Helm repo
  -h, --help                   help for init
  -p, --path string            path to the addon directory (default is ./<addon-name>)
      --version string         version of the Chart

Global Flags:
  -y, --yes   Assume yes for all user prompts
```
</details>

Examples:

**Normal creation**

Only when `./<addon-name>` (or user-specified path with `-p`) does not exist, or is an empty directory, will this command succeed. Otherwise, to prevent overwrites, this command will fail.

Screenshot from 2022-06-14 18-14-29

```console
$ vela addon create mongodb --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16 -p mongo/addon
Scaffold created in directory mongo/addon. What to do next:
- Review and edit what we have generated in mongo/addon
- Check out our guide on how to build your own addon: https://kubevela.io/docs/platform-engineers/addon/intro
- To enable the addon, run: vela addon enable mongo/addon
```

<details>
<summary>File Structure and Contents (Helm Component)</summary>

```console
$ vela addon init mongo-helm --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16
Scaffold created in directory mongo-helm. What to do next:
- Check out our guide on how to build your own addon: https://kubevela.io/docs/platform-engineers/addon/intro
- Review and edit what we have generated in mongo-helm
- To enable the addon, run: vela addon enable mongo-helm
```
```console
$ tree mongo-helm
mongo-helm
├── definitions
├── metadata.yaml
├── readme.md
├── resources
│   └── mongo-helm.cue
├── schemas
└── template.yaml

3 directories, 4 files
```
```console
$ cat mongo-helm/resources/mongo-helm.cue
output: {
	type: "helm"
	properties: {
		url:      "https://marketplace.azurecr.io/helm/v1/repo"
		repoType: "helm"
		chart:    "mongodb"
		version:  "12.1.16"
	}
}
```
```console
$ cat mongo-helm/metadata.yaml
dependencies:
- name: fluxcd
description: An addon for KubeVela.
icon: ""
invisible: false
name: mongo-helm
tags:
- 12.1.16
version: 12.1.16
```
```console
$ cat mongo-helm/readme.md
# Example mongo-helm Addon

Also check how to build your own addon: https://kubevela.net/docs/platform-engineers/addon/intro

## Directory Structure

- `template.yaml`: contains the basic app, you can add some component and workflow to meet your requirements. Other files in `resources/` and `definitions/` will be rendered as Components and appended in `spec.components`
- `metadata.yaml`: contains addon metadata information.
- `definitions/`: contains the X-Definition yaml/cue files. These file will be rendered as KubeVela Component in `template.yaml`
- `resources/`:
  - `parameter.cue` to expose parameters. It will be converted to JSON schema and rendered in UI forms.
  - All other files will be rendered as KubeVela Components. It can be one of the two types:
    - YAML file that contains only one resource. This will be rendered as a `raw` component
    - CUE template file that can read user input as `parameter.XXX` as defined `parameter.cue`.
      Basically the CUE template file will be combined with `parameter.cue` to render a resource.
      **You can specify the type and trait in this format**
```
```console
$ cat mongo-helm/template.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  creationTimestamp: null
  name: mongo-helm
  namespace: vela-system
spec:
  components: null
status: {}
```
</details>

<details>
<summary>File Structure and Contents (Basic)</summary>

```console
$ vela addon init mongo-sample
Scaffold created in directory mongo-sample. What to do next:
- Check out our guide on how to build your own addon: https://kubevela.io/docs/platform-engineers/addon/intro
- Review and edit what we have generated in mongo-sample
- To enable the addon, run: vela addon enable mongo-sample
```
```console
$ tree mongo-sample
mongo-sample
├── definitions
├── metadata.yaml
├── readme.md
├── resources
├── schemas
└── template.yaml

3 directories, 3 files
```
```console
$ cat mongo-sample/metadata.yaml
description: An addon for KubeVela.
icon: ""
invisible: false
name: mongo-sample
version: 1.0.0
```
```console
$ cat mongo-sample/template.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  creationTimestamp: null
  name: mongo-sample
  namespace: vela-system
spec:
  components: null
status: {}

```
</details>

**Directory already exists in local filesystem**
> If the directory is empty, it will behave the same as above.
```console
$ vela addon init mongodb --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16
Error: directory mongodb is not empty. To avoid any data loss, manually delete it first
```

**File with the same name already exists or other reasons**
```console
$ vela addon init main --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16
Error: we can't create directory main. Make sure the name has not already been taken and you have the proper rights to write to it
```

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Relevant unit tests have been added. Also manually tested several conditions described above.

### CODE REVIEW

**Comment by charlie0129** [comment]
> We can support a flag `-p` allow user to define the target path, default is `./`

Yes, manually specifying path is supported. Just `vela addon create ./some/path/to/addon-name`, `addon-name` will be used as addon name.

**Comment by wangyikewxgm** [comment]
> > We can support a flag `-p` allow user to define the target path, default is `./`
> 
> Yes, manually specifying path is supported. Just `vela addon create ./some/path/to/addon-name`, `addon-name` will be used as addon name.
> 
> https://github.com/kubevela/kubevela/blob/7bd5e06ff1249edcda03e698561012393eb09926/references/cli/addon.go#L379
IMO this is little bit tricky,


```
vela addon create name -p ./ 
```
Is better, and this is helm style.

File: pkg/addon/create.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+		return fmt.Errorf("cannot write %s: %w", templateFilePath, err)
+	}
+
+	// Write metadata.yaml with `fluxcd` as a dependency because we are using helm.
+	// However, this may change in the future, possibly with `argocd`.
+	metadataFilePath := path.Join(addonPath, MetadataFileName)
+	metadataTemplate := Meta{
+		Name:        addonName,
+		Version:     chartVersion,
+		Description: "Insert the description of your addon here.",
```
Comment (wangyikewxgm) [inline pkg/addon/create.go:196]: `Description: "An addon for KubeVela" ` ?

Comment (wangyikewxgm) [inline pkg/addon/create.go:196]: `Description: "An addon for KubeVela" ` ?

File: pkg/addon/create.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+		// Just use a dummy image with the addon name in it
+		Icon:         fmt.Sprintf("https://dummyimage.com/400x400/aaaaaa/333333&text=%s", addonName), // no need to escape url
+		URL:          "https://kubevela.io",
```
Comment (wangyikewxgm) [inline pkg/addon/create.go:199]: Set these fields empty by default?

Comment (wangyikewxgm) [inline pkg/addon/create.go:199]: Set these fields empty by default?

**Review by wangyikewxgm — COMMENTED** [review]
Generally LGTM. Good job! Please fix the comments.

File: references/cli/addon.go [hunk 3]
```diff
@@ -357,7 +358,64 @@ func NewAddonStatusCommand(c common.Args, ioStream cmdutil.IOStreams) *cobra.Com
...
+
+// NewAddonCreateCommand creates an addon scaffold
+func NewAddonCreateCommand() *cobra.Command {
+	var helmRepoURL string
+	var chartName string
+	var chartVersion string
+	var path string
+
+	cmd := &cobra.Command{
+		Use:   "create",
```
Comment (wonderflow) [inline references/cli/addon.go:373]:
```suggestion
		Use:   "scaffold",
```

Comment (wonderflow) [reply references/cli/addon.go:373]: or name it `init`

File: references/cli/addon.go [hunk 4]
```diff
@@ -357,7 +358,64 @@ func NewAddonStatusCommand(c common.Args, ioStream cmdutil.IOStreams) *cobra.Com
...
+			if len(args) < 1 {
+				return fmt.Errorf("an addon name is required")
+			}
+
+			return pkgaddon.CreateAddonFromHelmChart(args[0], path, helmRepoURL, chartName, chartVersion)
+		},
+	}
+
+	cmd.Flags().StringVar(&helmRepoURL, "helm-repo-url", "", "URL that points to a Helm repo")
+	err := cmd.MarkFlagRequired("helm-repo-url")
```
Comment (wonderflow) [inline references/cli/addon.go:400]: why these parameters are required? Is it possible for other kinds of type except helm? \cc @wangyikewxgm @charlie0129 

What should the `vela addon create/init` happen if we don't specify any helm parameters?  We **MUST** allow this case, at least we can create an empty scaffold with same basic desciptions/folders. Please refer to what helm do.

File: pkg/addon/create.go (hunk 1 above)
Comment (charlie0129) [reply pkg/addon/create.go]: changes made

File: pkg/addon/create.go (hunk 2 above)
Comment (charlie0129) [reply pkg/addon/create.go]: changes made

Comment (charlie0129) [reply pkg/addon/create.go]: changes made

File: references/cli/addon.go (hunk 3 above)
Comment (charlie0129) [reply references/cli/addon.go]: changes made

File: references/cli/addon.go (hunk 4 above)
Comment (charlie0129) [reply references/cli/addon.go:400]: If that's the case, then there won't be much to be generated, or too complicated to describe using parameters. Maybe just use the `example` addon in the `experimental` registry. That's probably a good start.

Comment (wonderflow) [reply references/cli/addon.go:400]: > Maybe just use the example addon in the experimental registry. 

Yes, that's somehow to the point. Make sure the first example can clarify all the concepts and rules of making an addon.

Comment (wangyikewxgm) [reply references/cli/addon.go:400]: > why these parameters are required? Is it possible for other kinds of type except helm? \cc @wangyikewxgm @charlie0129
> 
> What should the `vela addon create/init` happen if we don't specify any helm parameters? We **MUST** allow this case, at least we can create an empty scaffold with same basic desciptions/folders. Please refer to what helm do.

Helm parameter is optional. If without them will generate a basic structure.

File: pkg/addon/create.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+	// Currently, we do not check whether the Helm Chart actually exists, because it is just a scaffold.
+	// The user can still edit it after creation.
+	// Also, if the user is offline, we cannot check whether the Helm Chart exists.
+	// TODO(charlie0129): check whether the Helm Chart exists (if the user wants)
+
+	// Make sure url is valid
+	isValidURL := utils.IsValidURL(helmRepoURL)
+	if !isValidURL {
+		return fmt.Errorf("invalid helm repo url")
```
Comment (wonderflow) [inline pkg/addon/create.go:53]:
```suggestion
		return fmt.Errorf("invalid helm repo url %s", helmRepoURL)
```

File: references/cli/addon.go [hunk 6]
```diff
@@ -357,7 +358,69 @@ func NewAddonStatusCommand(c common.Args, ioStream cmdutil.IOStreams) *cobra.Com
...
+	├── resources
+	│   └── mongodb.cue
+	├── schemas
+	└── template.yaml
+
+If you want to store the scaffold in a different directory, you can use the -p/--path flag:
+	vela addon init mongodb -p ./some/repo --helm-repo-url=https://marketplace.azurecr.io/helm/v1/repo --chart=mongodb --version=12.1.16
+
+If you don't want the Helm component, just omit the three Chart-related parameters. We will create an empty scaffold for you.
+	vela addon init mongodb`,
```
Comment (wonderflow) [inline references/cli/addon.go:391]: The empty scaffold should also create other addon dir for convenience, it's better to put some example files in it.

https://kubevela.io/docs/platform-engineers/addon/intro#build-an-addon

Comment (charlie0129) [reply references/cli/addon.go:391]: Yes, there will be a minimal example in it (required files of an addon and other directories). 
<details>
<summary>File Structure and Contents (Basic)</summary>

```console
$ vela addon init mongo-sample
Scaffold created in directory mongo-sample. What to do next:
- Check out our guide on how to build your own addon: https://kubevela.io/docs/platform-engineers/addon/intro
- Review and edit what we have generated in mongo-sample
- To enable the addon, run: vela addon enable mongo-sample
```
```console
$ tree mongo-sample
mongo-sample
├── definitions
├── metadata.yaml
├── readme.md
├── resources
├── schemas
└── template.yaml

3 directories, 3 files
```
```console
$ cat mongo-sample/metadata.yaml
description: An addon for KubeVela.
icon: ""
invisible: false
name: mongo-sample
version: 1.0.0
```
```console
$ cat mongo-sample/readme.md
# mongo-sample

Also check how to build your own addon: https://kubevela.net/docs/platform-engineers/addon/intro

## Directory Structure

- `template.yaml`: contains the basic app, you can add some component and workflow to meet your requirements. Other files in `resources/` and `definitions/` will be rendered as Components and appended in `spec.components`
- `metadata.yaml`: contains addon metadata information.
- `definitions/`: contains the X-Definition yaml/cue files. These file will be rendered as KubeVela Component in `template.yaml`
- `resources/`:
  - `parameter.cue` to expose parameters. It will be converted to JSON schema and rendered in UI forms.
  - All other files will be rendered as KubeVela Components. It can be one of the two types:
    - YAML file that contains only one resource. This will be rendered as a `raw` component
    - CUE template file that can read user input as `parameter.XXX` as defined `parameter.cue`.
      Basically the CUE template file will be combined with `parameter.cue` to render a resource.
      **You can specify the type and trait in this format**
```
```console
$ cat mongo-sample/template.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  creationTimestamp: null
  name: mongo-sample
  namespace: vela-system
spec:
  components: null
status: {}

```
</details>

Comment (wonderflow) [reply references/cli/addon.go:391]: great

**Review by wonderflow — APPROVED** [review]

**Review by wangyikewxgm — APPROVED** [review]
great pr!

**Review by Somefive — APPROVED** [review]
Great Job!
